[code-infra] Add canary release scripts

michaldudak opened the pull request:
Created scripts in infra packages that can be used to publish them after each PR to the main (master or next) branch.

The workflow looks like this:
1. Check if there are any uncommitted changes. Fail if there are.
2. Set up .npmrc to look for an access token in an enviroment variable.
3. Figure out which packages changed between this and the previous commit.
4. Update package.json of the changed packages, adding the `-dev.[yyMMddHHmmss]` suffix.
5. Publish the changed packages.
6. Restore package.json files to the original state.

Runnning `pnpm canary:release` executes all these steps.
It requires an npm access token present in the NPM_TOKEN environment variable.

I'll create a GitHub action running the script in a separate PR, once this one is merged.

michaldudak commented:
I see. The script publishes changed packages and all the packages that depend on it, but "=> doesn't know about B@v2 because repo always resets versions after publishing" is indeed a problem.

> Only correct way is to always publish everything, changes or not

This will create a massive number of packages that are essentially the same (especially considering that we don't change the infra code very often).
How about we check for changes between the last commit and the last stable release instead?

Janpot commented:
> How about we check for changes between the last commit and the last stable release instead?

That could also work, if that's what we already do for stable releases anyway, I don't see an immediate problem with it. I'm adding it as an alternative to https://www.notion.so/mui-org/code-infra-Release-process-299a2fb7fe93441b8e345af0b2fc4fb4?pvs=4#7ba482d1725147bc8f312b6836f64b5e

michaldudak commented:
I ended up actually implementing both: first, the script checks if there are any changes in public packages between this and the previous commit (there is no point in releasing packages if nothing changed). Then we look for differences between this commit and the latest tag (or any other commit - this is configurable).

michaldudak commented:
I did a test run - the resulting packages were published to npm with the `-dev.240424162023-9968b4889d` suffix:
 - @mui/codemod
 - @mui/core-downloads-tracker
 - @mui/joy
 - @mui/material
 - @mui/docs
 - @mui/icons-material
 - @mui/lab
 - @mui/material-nextjs
 - @mui/styles
 - @mui/system

Janpot commented on scripts/canaryRelease.mts:
```diff
@@ -0,0 +1,242 @@
+
+async function setVersion(packages: PackageInfo[]) {
+  const { stdout: currentRevisionSha } = await $`git rev-parse --short HEAD`;
+  const timestamp = formatDate(new Date());
```
🤔 I was just thinking, we should be able to read the date from the commit. That way we make it impossible to publish the same commit twice.

oliviertassinari replied:
What's the issue with publishing the same commit twice?

Janpot replied:
To clarify, "Twice occasinally" is not really my main concern. It's about making this script idempotent. When multiple runs of the script has the same outcome. This makes it have an inherent resiliency which comes in handy when we're going to automate it. It becomes safe to do automatic retries, and an accidental misconfiguration can't make you end up with 100 times publishing the same package.

michaldudak replied:
Good point!

oliviertassinari commented on scripts/canaryRelease.mts:
```diff
@@ -0,0 +1,242 @@
+        newVersion = version.slice(0, dashIndex);
+      }
+
+      newVersion = `${newVersion}-dev.${timestamp}-${currentRevisionSha}`;
```
In the format, is it possible to drop the time from the date? https://www.npmjs.com/package/typescript?activeTab=versions seems clearer and good enough since we have the commit.

oliviertassinari replied:
Actually, maybe even?

6.0.0-dev-2024-04-24-9968b4889d

would be clearer.

michaldudak replied:
This was done to ensure the version suffix is always increasing. In the TypeScript case, it's OK, as they publish one version a day. If we're targeting to publish each commit, having time is required to ensure increasing versions.

oliviertassinari commented on scripts/canaryRelease.mts:
```diff
@@ -0,0 +1,242 @@
+  return date
+    .toISOString()
+    .replace(/[-:TZ]/g, '')
+    .slice(2, 14);
```
Why slice the year? It feels like it breaks the pattern matching for what the string is about.

```suggestion
    .slice(0, 14);
```

michaldudak commented:
Updated the version number to reflect the commit date and changed the format to `-dev.yyyyMMdd-hhmmss-sha` (for example `@mui/material@6.0.0-dev.20240507-080616-4ea4f9e382`)

Janpot reviewed (approved):
Good for me 👍

michaldudak commented:
One more addition I made - the ability to ignore certain packages. I thought it could be useful to prevent uploading the huge @mui/icons-material after each commit.
